soc/arm/microchip_mec/mec172x/soc.c

```diff
@@ -0,0 +1,411 @@
+	for (n = 0; n < MCHP_GIRQ_IDX_MAX; n++) {
+		eciar->GIRQ[n].EN_CLR = 0xFFFFFFFFul;
+	}
+
+	/* Clear all external NVIC enables and pending status */
+	for (n = 0u; n < MCHP_NUM_NVIC_REGS; n++) {
+		NVIC->ICER[n] = 0xFFFFFFFFul;
+		NVIC->ICPR[n] = 0xFFFFFFFFul;
+	}
+
+	return 0;
+}
+
+static bool is_sil_osc_enabled(void)
+{
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+
+	if (vbr->CLK32_SRC & MCHP_VBATR_CS_SO_EN) {
+		return true;
+	}
+
+	return false;
+}
+
+static void enable_sil_osc(uint32_t delay_cnt)
+{
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_SO_EN;
+	spin_delay(delay_cnt);
+}
+
+static int enable_32k_pin(uint32_t wait_cnt)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+	struct gpio_ctrl_regs *gpcr = (struct gpio_ctrl_regs *)GPIO_BASE;
+
+	gpcr->CTRL_0165 = PIN_32KHZ_IN_MODE;
+
+	do {
+		if (pcr->PWR_RST_STS & MCHP_PCR_PRS_32K_ACTIVE_RO) {
+			pcr->CLK32K_SRC_VTR = MCHP_PCR_VTR_32K_SRC_SILOSC;
+			return 0;
+		}
+	} while (wait_cnt--);
+
+	return -1;
+}
+
+
+/* Do nothing if crystal configuration matches request. */
+static int enable_32k_crystal(uint32_t flags)
+{
+	struct vbatr_regs *vbr = VBATR_XEC_REG_BASE;
+	uint32_t vbcs = vbr->CLK32_SRC;
+	uint32_t cfg = MCHP_VBATR_CS_SO_EN;
+
+	if (flags & CLK32K_FLAG_CRYSTAL_SE) {
+		cfg |= MCHP_VBATR_CS_XTAL_SE;
+	}
+
+	if ((vbcs & cfg) == cfg) {
+		return 0;
+	}
+
+	/* parallel and clear high startup current disable */
+	vbcs &= ~(MCHP_VBATR_CS_SO_EN | MCHP_VBATR_CS_XTAL_SE |
+		  MCHP_VBATR_CS_XTAL_DHC | MCHP_VBATR_CS_XTAL_CNTR_MSK);
+	if (flags & CLK32K_FLAG_CRYSTAL_SE) { /* single ended connection? */
+		vbcs |= MCHP_VBATR_CS_XTAL_SE;
+	}
+
+	/* XTAL mode and enable must be separate writes */
+	vbr->CLK32_SRC = vbcs | MCHP_VBATR_CS_XTAL_CNTR_DG;
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_SO_EN;
+	spin_delay(CLK32K_XTAL_WAIT);
+	vbr->CLK32_SRC |= MCHP_VBATR_CS_XTAL_DHC;
+
+	return 0;
+}
+
+/*
+ * Counter checks:
+ * 32KHz period counter minimum for pass/fail: 16-bit
+ * 32KHz period counter maximum for pass/fail: 16-bit
+ * 32KHz duty cycle variation max for pass/fail: 16-bit
+ * 32KHz valid count minimum: 8-bit
+ *
+ * 32768 Hz period is 30.518 us
+ * HW count resolution is 48 MHz.
+ * One 32KHz clock pulse = 1464.84 48 MHz counts.
+ */
+#define CNT32K_TMIN 1435
+#define CNT32K_TMAX 1495
+#define CNT32K_DUTY_MAX 74
+#define CNT32K_VAL_MIN 4
+
+static int check_32k_crystal(void)
+{
+	struct pcr_regs *pcr = PCR_XEC_REG_BASE;
+
+	pcr->CNT32K_CTRL = 0U;
+	pcr->CLK32K_MON_IEN = 0U;
+	pcr->CLK32K_MON_ISTS = MCHP_PCR_CLK32M_ISTS_MASK;
+
+	pcr->CNT32K_PER_MIN = CNT32K_TMIN;
+	pcr->CNT32K_PER_MAX = CNT32K_TMAX;
+	pcr->CNT32K_DV_MAX = CNT32K_DUTY_MAX;
+	pcr->CNT32K_VALID_MIN = CNT32K_VAL_MIN;
+
+	pcr->CNT32K_CTRL =
+		MCHP_PCR_CLK32M_CTRL_PER_EN | MCHP_PCR_CLK32M_CTRL_DC_EN |
+		MCHP_PCR_CLK32M_CTRL_VAL_EN | MCHP_PCR_CLK32M_CTRL_CLR_CNT;
+
+	uint32_t status = pcr->CLK32K_MON_ISTS;
+
+	while (status == 0) {
+		status = pcr->CLK32K_MON_ISTS;
```

Comment:
Prefer to have another break condition (may be like count or timeout) to avoid infinite loop incase of sticky status

Reply:
Hi @VenkatKotakonda  It's done.